# Setup: Install Required Packages

In [17]:
import requests

# Part 1: Working with JSON Data

In [18]:
sample_json = '''
{
  "station": "USC00305800",
  "name": "New York Central Park",
  "location": {
    "latitude": 40.7789,
    "longitude": -73.9692
  },
  "observations": [
    {"date": "2023-01-01", "temperature": 32, "precipitation": 0.0},
    {"date": "2023-01-02", "temperature": 28, "precipitation": 0.5},
    {"date": "2023-01-03", "temperature": 35, "precipitation": 0.0},
    {"date": "2023-01-04", "temperature": 38, "precipitation": 0.2},
    {"date": "2023-01-05", "temperature": 41, "precipitation": 0.0}
  ]
}
'''

In [19]:
data = json.loads(sample_json)

In [20]:
print("Station:", data['station'])

Station: USC00305800


In [21]:
print("Location:", data['location'])

Location: {'latitude': 40.7789, 'longitude': -73.9692}


In [22]:
print("First observation:", data['observations'][0])

First observation: {'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}


## Task 1

In [23]:
print("Date, Temperature")
for obs in data['observations']: print(obs['date'], obs['temperature'])

Date, Temperature
2023-01-01 32
2023-01-02 28
2023-01-03 35
2023-01-04 38
2023-01-05 41


## Task 2

In [24]:
total_temp = 0
count = 0

In [25]:
for obs in data['observations']:
    total_temp += obs['temperature']
    count += 1

In [26]:
average_temp = total_temp / count
print(f"The average temperature is: {average_temp}")

The average temperature is: 34.8


## Task 3

In [27]:
print("\nDays with precipitation:")


Days with precipitation:


In [28]:
for obs in data['observations']:
    if obs['precipitation'] > 0:
            print(f"Date: {obs['date']}, Precipitation: {obs['precipitation']}")

Date: 2023-01-02, Precipitation: 0.5
Date: 2023-01-04, Precipitation: 0.2


# Part 2: Downloading Files with Python

In [30]:
%pip install pooch

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [33]:
import pooch

In [34]:
file_path = pooch.retrieve(
    url="https://github.com/pandas-dev/pandas/raw/main/doc/data/air_quality_no2.csv",
    known_hash=None
)

print("File downloaded to:", file_path)
print("File exists:", os.path.exists(file_path))

File downloaded to: /home/eva2122/.cache/pooch/458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
File exists: True


## Task 1

In [35]:
import os

In [36]:
with open(file_path, 'r') as f:
    for i in range(5):
        print(f.readline())

datetime,station_antwerp,station_paris,station_london

2019-05-07 02:00:00,,,23.0

2019-05-07 03:00:00,50.5,25.0,19.0

2019-05-07 04:00:00,45.0,27.7,19.0

2019-05-07 05:00:00,,50.4,16.0



In [37]:
file_size = os.path.getsize(file_path)
print(f"File size: {file_size} bytes")

File size: 31984 bytes


In [38]:
with open(file_path, 'r') as f:
    line_count = len(f.readlines())

print(f"The file has {line_count} lines.")

The file has 1036 lines.


## Task 2

In [29]:
print("Getting real weather data")
url = "https://api.weather.gov/points/40.7789,-73.9692"
response = requests.get(url)
if response.status_code == 200:
    weather_data = response.json()
    city = weather_data['properties']['relativeLocation']['properties']['city']
    state = weather_data['properties']['relativeLocation']['properties']['state']
    print(f"Location: {city}, {state}")

    forecast_url = weather_data['properties']['forecast']
    forecast_response = requests.get(forecast_url)
    forecast = forecast_response.json()

    print("\nForecast:")
    print("-" * 50)
    for period in forecast['properties']['periods'][:3]:
        print(f"{period['name']}:")
        print(f" Temperature: {period['temperature']}°{period['temperatureUnit']}")
        print(f" Forecast: {period['shortForecast']}")

else:
    print(f"Error: {response.status_code}")

Getting real weather data
Location: Guttenberg, NJ

Forecast:
--------------------------------------------------
This Afternoon:
 Temperature: 45°F
 Forecast: Mostly Sunny
Tonight:
 Temperature: 38°F
 Forecast: Mostly Cloudy
Wednesday:
 Temperature: 42°F
 Forecast: Light Rain Likely


In [39]:
with open(file_path, 'r') as f:
    line_count = len(f.readlines())

print(f"The file has {line_count} lines.")

The file has 1036 lines.


## Task 3

In [42]:
print("\nData Inventory:")
print("1. air_quality_no2.csv - Air quality NO2 measurements")


Data Inventory:
1. air_quality_no2.csv - Air quality NO2 measurements


# Part 3: Understanding NetCDF Metadata

In [47]:
print("Getting dataset information...")
base_url = "http://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"
dds_url = base_url + ".dds"
response = requests.get(dds_url)

print("dataset Structure (DDS):")
print("=" * 60)
print(response.text[:500])
print("=" * 60)

Getting dataset information...
dataset Structure (DDS):
Dataset {
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Float32 T[T = 324];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



## Task 1

The dimensions are T (time) 324, Y (latitude) 360, x (longitude) 720.

The main variable is rain.

## Task 2

In [50]:
print("Getting dataset attributes...")
das_url = base_url + ".das"
resposne = requests.get(das_url)
print("Dataset Attributes (DAS):")
print("=" * 60)
print(response.text[:1000])
print("=" * 60)

Getting dataset attributes...
Dataset Attributes (DAS):
Dataset {
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Float32 T[T = 324];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



## Task 3

The dataset contains global monthly rainfall measurements.
The data is measured in months since January 1960 and has 27 years of data.
Y (latitude) 360 points degrees north.
X (longitude 720 points degrees east.
Precipitation measured millimeters per day.